### PR TITLE
BUG: special: fix hyp2f1 behavior in certain circumstances

### LINKED ISSUE #7961: Wrong value for hyp2f1(k, -n, -n, z) with k,n positive integers
The following values are wrong
```
from scipy.special import hyp2f1
print(hyp2f1(2, -1, -1, 0.7)) # -> 11.1111111111  (should be 2.4)
print(hyp2f1(2, -2, -2, 0.7)) # -> 11.1111111111  (should be 3.87)
```
These were previously calculated wrong also by mpmath, but that's fixed in mpmath-1.0.0.
As a consequence, the test `scipy.special.tests.test_mpmath::test_hyp2f1_strange_points` fails with the fixed mpmath (marked as xfail in gh-7962).

### Scipy/Numpy/Python version information:
```
1.0.0rc1 1.13.1 sys.version_info(major=3, minor=6, micro=2, releaselevel='final', serial=0)
```

### PR DESCRIPTION
**Pull request by ghost**
Hyp2f1 previously failed when c ==b and was a negative integer greater than -100. This resolves the behavior by implementing a solution given on stackexchange to fix the behavior so that it aligns with what people expect.

I broke out the functions for maximum clarity. Maybe the compiler can optimize them away?

Closes #7961.

### CODE REVIEW

**Comment by xoviat** [comment]
Also, I'm more than happy to accept feedback to improve performance.

File: scipy/special/cephes/hyp2f1.c [hunk 1]
```diff
@@ -83,6 +83,43 @@ static double hyt2f1(double a, double b, double c, double x, double *loss);
...
+static double pochhammer(double x, double n);
+static double factorial(double k);
+static double hyp2f1_indeterminate(double a, double b, double x);
+
+/*
+    See http://mathworld.wolfram.com/PochhammerSymbol.html
+*/
+static double pochhammer(double x, double n)
+{
+    return gamma(x + n) / gamma(x);
```
Comment (pv) [inline scipy/special/cephes/hyp2f1.c]: Use the one defined in scipy/scipy/special/c_misc/poch.c (it's available in the compilation unit, just needs prototype).
Naive computation overflows very easily.

File: scipy/special/cephes/hyp2f1.c [hunk 2]
```diff
@@ -83,6 +83,43 @@ static double hyt2f1(double a, double b, double c, double x, double *loss);
...
+*/
+static double factorial(double k)
+{
+    return gamma(k + 1);
+}
+
+/*
+    See https://math.stackexchange.com/questions/2457831/hypergeometric-expansion-approximation
+*/
+static double hyp2f1_indeterminate(double a, double b, double x)
```
Comment (pv) [inline scipy/special/cephes/hyp2f1.c]: Move all helper functions after the main function, like the other special cases.

Also, more specific function name could be better (regardless of the fact that the other functions follow fortran 66 naming style).

Please also prefer citing an appropriate DLMF or abramowitz&stegun formula over stackoverflow posts.

File: scipy/special/cephes/hyp2f1.c [hunk 3]
```diff
@@ -83,6 +83,43 @@ static double hyt2f1(double a, double b, double c, double x, double *loss);
...
+/*
+    See http://mathworld.wolfram.com/PochhammerSymbol.html
+*/
+static double pochhammer(double x, double n)
+{
+    return gamma(x + n) / gamma(x);	
+}
+
+/*
+    See http://mathworld.wolfram.com/GammaFunction.html
```
Comment (pv) [inline scipy/special/cephes/hyp2f1.c]: No need for separate function, please write in terms of gamma.

File: scipy/special/cephes/hyp2f1.c [hunk 4]
```diff
@@ -130,7 +167,15 @@ double a, b, c, x;
     if (ax < 1.0 || x == -1.0) {
 	/* 2F1(a,b;b;x) = (1-x)**(-a) */
 	if (fabs(b - c) < EPS) {	/* b = c */
-	    y = pow(s, -a);	/* s to the -a power */
+	    if (b < 0 && b > -100 && (floor(b) - b) < EPS) {
```
Comment (pv) [inline scipy/special/cephes/hyp2f1.c]: AKA `if (neg_int_b && b > -100)`

File: scipy/special/cephes/hyp2f1.c [hunk 5]
```diff
@@ -130,7 +167,15 @@ double a, b, c, x;
...
+		/*
+		    The approximation becomes good enough 
+		    again once we go below -100
```
Comment (pv) [inline scipy/special/cephes/hyp2f1.c]: It's not an approximation --- the formula is exact. The point is that when c is integer <= 0, the result is not uniquely defined by continuity.

Also, I think the statement is not correct? For example, if `x > 1`, the equation below produces nan but the sum still terminates.

File: scipy/special/cephes/hyp2f1.c [hunk 6]
```diff
@@ -83,6 +83,43 @@ static double hyt2f1(double a, double b, double c, double x, double *loss);
...
+static double hyp2f1_indeterminate(double a, double b, double x)
+{
+    /*
+    	"converting floating point to integers is shockingly expensive"
+    */
+    double k;
+    double sum = 0;
+
+    for (k = 0; k <= -b; k++) {
+	sum += pochhammer(a, k) * pow(x, k) / factorial(k);
```
Comment (pv) [inline scipy/special/cephes/hyp2f1.c]: Ideally, you would also calculate the error estimate for the sum (relevant for negative `x` when it is alternating) and return nan on total precison loss.
Unfortunately, the rest of hyp2f1 is not written as carefully.

**Comment by xoviat** [comment]
Unmarking as WIP because we get an XPASS from the test.

File: scipy/special/cephes/hyp2f1.c (hunk 1 above)
Comment (xoviat) [reply scipy/special/cephes/hyp2f1.c]: Will do.

File: scipy/special/cephes/hyp2f1.c (hunk 3 above)
Comment (xoviat) [reply scipy/special/cephes/hyp2f1.c]: Will do.

File: scipy/special/cephes/hyp2f1.c (hunk 2 above)
Comment (xoviat) [reply scipy/special/cephes/hyp2f1.c]: Give me a name for the function. I'm terrible at names.

File: scipy/special/cephes/hyp2f1.c (hunk 5 above)
Comment (xoviat) [reply scipy/special/cephes/hyp2f1.c]: `if ax < 1.0`

File: scipy/special/cephes/hyp2f1.c (hunk 4 above)
Comment (xoviat) [reply scipy/special/cephes/hyp2f1.c]: Will do.

**Comment by xoviat** [comment]
@pv What should the function name be?

**Comment by xoviat** [comment]
I cannot find a reference to that formula other than the stackexchange post.

**Comment by xoviat** [comment]
I have (hopefully!) implemented the collector pattern correctly.

**Comment by pv** [comment]
It is 15.4.2 in Abramowitz & Stegun.

**Comment by pv** [comment]
Now that you are editing it, please extend the mpmath comparison test to cover very large, small, and also negative values of `a` and `x`.

File: scipy/special/tests/test_mpmath.py [hunk 7]
```diff
@@ -110,6 +110,13 @@ def test_hyp2f1_strange_points():
     pts = [
         (2, -1, -1, 0.7),  # expected: 2.4
         (2, -2, -2, 0.7),  # expected: 3.87
+        (1000, -2, -2, 0.7),
+        (2, -2, -2, 100),
+        (-5, -2, -2, 100),
+        (-100, -2, -2, 0.7),
+        (-100, -2, -2, -0.5),
+        (-100, -2, -2, -50),
+        (105, -2, -2, -50),
```
Comment (pv) [inline scipy/special/tests/test_mpmath.py]: Also extreme stuff like non-integers,  large numbers `1e30`, larger `|b|=|c|` etc is useful to check --- it's the corner cases that are hard. You can probably use itertools.product etc. to generate combinations to avoid making long lists.

File: scipy/special/cephes/hyp2f1.c [hunk 8]
```diff
@@ -590,3 +601,30 @@ static double hyp2f1ra(double a, double b, double c, double x,
...
+    double k;
+    double sum = 0;
+    double x_k = 1; // x^k
+    double k_f = 1; // k!    
+
+    for (k = 0; k <= -b; k++) {
+	if (k != 0) {
+	    k_f *= k;  // We have a single CPU pipeline loss (not bad!)
+	}
+	sum += poch(a, k) * x_k / k_f;
```
Comment (person142) [inline scipy/special/cephes/hyp2f1.c]: The factorial will overflow quickly so you don't want to explicitly compute it. Instead have a variable like `term` which contains the current term and then gets updated at every step by multiplying by `(a + k)*x/k` (probably some index errors there but you get the idea).

File: scipy/special/cephes/hyp2f1.c [hunk 9]
```diff
@@ -590,3 +601,30 @@ static double hyp2f1ra(double a, double b, double c, double x,
...
     return f0;
 }
+
+
+/*
+    15.4.2 Abramowitz & Stegun.
+*/
+static double hyp2f1_neg_c_equal_bc(double a, double b, double x)
+{
+    /*
```
Comment (person142) [inline scipy/special/cephes/hyp2f1.c]: You don't need these comments.

File: scipy/special/cephes/hyp2f1.c [hunk 10]
```diff
@@ -590,3 +601,30 @@ static double hyp2f1ra(double a, double b, double c, double x,
...
+    	"converting floating point to integers is shockingly expensive"
+	
+	Therefore, we stay with double the entire time.
+    */
+    double k;
+    double sum = 0;
+    double x_k = 1; // x^k
+    double k_f = 1; // k!    
+
+    for (k = 0; k <= -b; k++) {
```
Comment (person142) [inline scipy/special/cephes/hyp2f1.c]: Note that this is going to involve summing a very large series which could be inaccurate (when there's cancellation). Probably need to investigate when it will give bad results and return `nan` in those cases.

**Review by person142 — CHANGES_REQUESTED** [review]

File: scipy/special/cephes/hyp2f1.c (hunk 10 above)
Comment (xoviat) [reply scipy/special/cephes/hyp2f1.c]: No. The series is only summed to 100 at most.

**Comment by person142** [comment]
I see that @pv already had some of the same comments as me, apologies for any repeats.

File: scipy/special/cephes/hyp2f1.c (hunk 8 above)
Comment (xoviat) [reply scipy/special/cephes/hyp2f1.c]: As I said, `k < 100` in the worst-case scenario.

Comment (xoviat) [reply scipy/special/cephes/hyp2f1.c]: Wait, that's me being idiotic.

File: scipy/special/cephes/hyp2f1.c (hunk 10 above)
Comment (pv) [reply scipy/special/cephes/hyp2f1.c]: However, the statement that the different definition of the `c=b=negint` case gives the same result for `b<-100` is questionable, consider e.g. `hyp2f1(10, -101, -101, 0.9)`

File: scipy/special/cephes/hyp2f1.c [hunk 11]
```diff
@@ -590,3 +601,23 @@ static double hyp2f1ra(double a, double b, double c, double x,
 
     return f0;
 }
+
+
+/*
+    15.4.2 Abramowitz & Stegun.
+*/
+static double hyp2f1_neg_c_equal_bc(double a, double b, double x)
+{
```
Comment (person142) [inline scipy/special/cephes/hyp2f1.c]: This should be (untested):

```
double k;
double collector = 1;
double sum = 1;

for (k = 1; k <= -b; k++) {
    collector *= (a + k - 1)*x/k;
    sum += collector;
}
```

**Comment by xoviat** [comment]
Before I proceed, I would like to note that this comment by @pv is correct:

> This sort of statements should be accompanied by reasoning why it is so, otherwise it's very easy to make mistakes.

Unfortunately, because I am not very strong in this area, it's going to take me some time to work out the behavior of the function to make sure that I'm not missing anything important. Of course I would not mind someone else working this out if they have time.

File: scipy/special/cephes/hyp2f1.c (hunk 11 above)
Comment (xoviat) [reply scipy/special/cephes/hyp2f1.c]: Yeah, the problem was that I did not expand the `poch` function function so I did not expand it.

File: scipy/special/cephes/hyp2f1.c [hunk 12]
```diff
@@ -130,7 +132,16 @@ double a, b, c, x;
     if (ax < 1.0 || x == -1.0) {
 	/* 2F1(a,b;b;x) = (1-x)**(-a) */
 	if (fabs(b - c) < EPS) {	/* b = c */
-	    y = pow(s, -a);	/* s to the -a power */
+	    if (neg_int_b && b > -100) {
```
Comment (xoviat) [inline scipy/special/cephes/hyp2f1.c]: TODO: prove the `-100`.

**Comment by person142** [comment]
> it's going to take me some time to work out the behavior of the function to make sure that I'm not missing anything important

Probably the safe thing to do for now is to always use the series and follow @pv's advice to track the error being accumulated and then bail out with `nan` if it is too large. This corrects some of the problematic cases and doesn't return wrong/uncertain results in the other ones.

Further improvements can always be made later. :-)

**Comment by pv** [comment]
... and an estimate for the relative error in the sum is 
`1e-16 * (1 + largest_term_fabs/fabs(sum))`. We typically start
returning nan if it gets worse than 1e-7 (ie sqrt eps).

**Comment by xoviat** [comment]
@pv The error is checked at the end to avoid a performance cost (checking again and again inside of the loop).

File: scipy/special/cephes/hyp2f1.c [hunk 13]
```diff
@@ -590,3 +595,27 @@ static double hyp2f1ra(double a, double b, double c, double x,
...
+    double err;
+    double collector = 1;
+    double sum = 1;
+
+    for (k = 1; k <= -b; k++) {
+  	collector *= (a + k - 1)*x/k;
+    	sum += collector;
+    }
+
+    err = 1e-16 * (1 + fabs(collector)/fabs(sum));
```
Comment (pv) [inline scipy/special/cephes/hyp2f1.c]: I think you need to do `if (fabs(collector) > max_collector) max_collector = collector;` inside the loop, unless you can prove the largest term is found at the end (not obvious).

Comment (xoviat) [reply scipy/special/cephes/hyp2f1.c]: I thought about this for a minute to make sure that I was (hopefully correct). But here goes:

1. Ignore "x" because it is constant throughout the loop.
2. Bracket "a" from -1 to 1 from the above program conditions
3. Simplify the expression: the numerator is from "1 + k - 1" = "k" to "-1 + k - 1" = "k - 2"
4. See that in the first case, it's "k/k"= constant (first term = last term)
5. See that in the second case, it's "(k - 2)/k", which has a limit behavior of 1 (starts smaller and grows).

In summary, it seems to start out small and then approach 1 in all cases, just from different starting places.

Comment (xoviat) [reply scipy/special/cephes/hyp2f1.c]: Also, I don't think that the series is alternating.

Comment (pv) [reply scipy/special/cephes/hyp2f1.c]: For `x < 0` the series is alternating.

Step 1. above sounds problematic. Better to move the check inside the loop, so we don't need to reason about this.

Comment (xoviat) [reply scipy/special/cephes/hyp2f1.c]: Even if it is alternating, it would still tend to approach 1/-1.

Comment (xoviat) [reply scipy/special/cephes/hyp2f1.c]: It's ultimately a (k - 2)/k, (k - 1.5)/k, (k - 1)/k, k/k problem. And we're comparing the absolute value, so the negative sign can be dropped.

File: scipy/special/tests/test_mpmath.py [hunk 14]
```diff
@@ -107,10 +108,8 @@ def test_hyp0f1_gh_1609():
 @pytest.mark.xfail(reason="hyp2f1 produces wrong/nonstandard values (gh-7961)")
 @check_version(mpmath, '1.0.0')
 def test_hyp2f1_strange_points():
-    pts = [
-        (2, -1, -1, 0.7),  # expected: 2.4
-        (2, -2, -2, 0.7),  # expected: 3.87
-    ]
+    numbers = [2, -2, -1, 0.7, 100, -50, 1e30, -1000]    
+    pts = itertools.product(numbers, repeat=4)
     kw = dict(eliminate=True)
```
Comment (xoviat) [inline scipy/special/tests/test_mpmath.py]: Based on what happened on travis, this may be too much.

Comment (xoviat) [reply scipy/special/tests/test_mpmath.py]: Possibly it triggers an effectively infinite for loop?

**Comment by pv** [comment]
The point is floating point loss of precision when summing alternating 
series. I've explained how to do it in a way that's certainly correct, 
and don't really want to start reviewing alternatives --- it is doubtful 
the performance differences matter much in practice, as long as you get 
the big-O right.

**Comment by pv** [comment]
Exercise:
```
import mpmath, numpy as np
from scipy.special import poch, gamma

a = -12.5
b = -60
x = 0.9

k = np.arange(0, -b + 1)
terms = poch(a, k) * x**k / gamma(1 + k)
s = terms.sum()
maxterm = abs(terms).max()

mpmath.mp.dps = 100
s_mp = float(mpmath.hyp2f1(a, b, b, x))

print("rel. error:", abs(s - s_mp)/abs(s_mp))
print("rel. error estimate:", np.finfo(float).eps * (1 + maxterm / s))
print("last term error est:", np.finfo(float).eps * (1 + abs(terms[-1]) 
/ s))
```

**Comment by xoviat** [comment]
> a = -12.5

I misread `ax  < 1.0` and bracketed the wrong variable (`a` rather than `x`). Silly me.

**Comment by xoviat** [comment]
Reopened to get @person142's build ahead of me.

File: scipy/special/cephes/hyp2f1.c [hunk 15]
```diff
@@ -590,3 +595,27 @@ static double hyp2f1ra(double a, double b, double c, double x,
...
+    double err;
+    double collector = 1;
+    double sum = 1;
+    double const max_err = (1e-7/1e-16) - 1;
+
+    for (k = 1; k <= -b; k++) {
+  	collector *= (a + k - 1)*x/k;
+    	sum += collector;
+    
+        if (fabs(collector / sum) > max_err) {
```
Comment (pv) [inline scipy/special/cephes/hyp2f1.c]: Please change the error estimate to the form I suggested earlier, so that it's not necessary to make assumptions about the behavior of the sum.

File: scipy/special/cephes/hyp2f1.c [hunk 16]
```diff
@@ -590,3 +595,27 @@ static double hyp2f1ra(double a, double b, double c, double x,
...
+*/
+static double hyp2f1_neg_c_equal_bc(double a, double b, double x)
+{
+    double k;
+    double err;
+    double collector = 1;
+    double sum = 1;
+    double const max_err = (1e-7/1e-16) - 1;
+
+    for (k = 1; k <= -b; k++) {
```
Comment (pv) [inline scipy/special/cephes/hyp2f1.c]: This needs a check along the lines of 
`if (!(fabs(b) < 1e15)) { return NPY_NAN; }`, otherwise it enters an infinite loop when you run out bits in the mantissa.

File: scipy/special/cephes/hyp2f1.c (hunk 15 above)
Comment (xoviat) [reply scipy/special/cephes/hyp2f1.c]: I assume that I am wrong here, but I would like to know why: I literally just moved all of the constant parts to one side of the equation and expanded the fabs around the variable expression.

Comment (pv) [reply scipy/special/cephes/hyp2f1.c]: This is comparing the current fp error to the current partial sum. This is not necessarily the same as the maximum fp error vs. the final value of the sum.

Consider for example the terms `[1e15, -1e15, 1e30]`. There's no problem computing the value of the sum, but the above if statement bails out. Ruling out that this sort of thing cannot occur in the above sum requires some mathematical argument, and it's simpler to not try doing that.

File: scipy/special/tests/test_mpmath.py [hunk 17]
```diff
@@ -111,6 +112,7 @@ def test_hyp2f1_strange_points():
         (2, -1, -1, 0.7),  # expected: 2.4
```
Comment (pv) [inline scipy/special/tests/test_mpmath.py]: As a final change, I think we can remove the xfail mark above.

Comment (xoviat) [reply scipy/special/tests/test_mpmath.py]: Aside from the MSVC failures (which I can fix later), the `hpy2f1` test didn't pass with the additional points. I can revert changes and then unmark the test as xfail, but there's probably an additional failure somewhere that will need a whole new round of investigation :(.

File: scipy/special/tests/test_mpmath.py [hunk 18]
```diff
@@ -111,6 +112,7 @@ def test_hyp2f1_strange_points():
...
         (2, -2, -2, 0.7),  # expected: 3.87
     ]
+    pts += list(itertools.product([2, 1, -0.7, -1000], repeat=4))
```
Comment (pv) [inline scipy/special/tests/test_mpmath.py]: I think we should here limit to the cases where `b==c==negative integer`.

**Comment by xoviat** [comment]
@pv Concerns have been addressed and I think the CI will pass. Would you like me to clean up the commits?

File: scipy/special/cephes/hyp2f1.c [hunk 19]
```diff
@@ -590,3 +596,32 @@ static double hyp2f1ra(double a, double b, double c, double x,
...
+    double sum = 1;
+    double collector_max = 1;
+
+    if (!(fabs(b) < 1e5)) {
+        return NPY_NAN;
+    }
+
+    for (k = 1; k <= -b; k++) {
+        collector *= (a + k - 1)*x/k;
+        collector_max = fmax(fabs(collector), collector_max);
```
Comment (person142) [inline scipy/special/cephes/hyp2f1.c]: Is the complicated `fmax` logic really necessary here?

Comment (xoviat) [reply scipy/special/cephes/hyp2f1.c]: Yes. `fmax` avoids branching instructions for all compilers except MSVC 2010 and MSVC 2008 (and I don't really care that Python 2.7 and Python 3.4 on windows will run slower).

Comment (xoviat) [reply scipy/special/cephes/hyp2f1.c]: In C, it looks more complicated (well, if you're not used to functional programming). But in assembly, it's considerably simpler.

**Comment by xoviat** [comment]
@pv Does this still need more work?

File: scipy/special/cephes/hyp2f1.c (hunk 19 above)
Comment (person142) [reply scipy/special/cephes/hyp2f1.c]: In that case the `fmax`/`fmin` code should probably go in `special/_c99compat.h` where the other C99 functions are.

**Comment by pv** [comment]
The fmax.h code is needed now in ~three different places in scipy.
Maybe should be combined to a single file and just incldue paths adjusted.

**Comment by person142** [comment]
Are you thinking ~three places in just special or in multiple submodules?

**Comment by pv** [comment]
$ find scipy -name 'fmax.h'
scipy/spatial/ckdtree/src/fmax.h
scipy/optimize/_trlib/trlib/fmax.h

**Comment by person142** [comment]
And now that I look there's code for setting the fpu mode in `_lib`, which is another C99 feature that was (in a roundabout way) recreated in special for `wrightomega`. So gathering C99 stuff in `_lib` sounds like a great idea.

**Comment by xoviat** [comment]
`fmax.h` duplication addressed in #8003. You can merge that one first and then I can rebase and fix the duplication in this PR or you can merge this one first and I can fix the duplication in that PR.

**Comment by xoviat** [comment]
Unfortunately MSVC 2008 isn't happy. The linker error points to an unresolved external symbol, but `fmax` should be `static`, not `extern`.

**Comment by xoviat** [comment]
Wow, that was just nasty. I picked up another `_c99compat.h` header.

**Comment by xoviat** [comment]
That was more than I bargained for.

@pv Please merge #8006 first and then I will rebase for a clean patch.

**Comment by pv** [comment]
thanks, merged
